cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+
+### Resources
+
+TODO: This section needs an update.
+
+#### ConfigMap - cilium-config
+
+* option-name: `default`
+  * description
+* mesh-auth-mode: `disabled`
+  * enabled / disabled
+* mesh-auth-cert-source: `spiffe`
+  * Configure certificate mechanism, default to directory path
+* mesh-spire-server: `____`
+  * Configure the spire server
+* mesh-spire-agent: `unix://var/run/cilium/spiffe/agent.sock`
+  * Configure the SPIRE agent
+* mesh-auth-path: `/var/run/cilium/auth/`
+  * Configure the directory to pull certificates from, if mesh-auth-cert is directory
+* mesh-auth-port: `4250`
+  * Each agent opens this port for mutual authentication
+  * TODO review port for conflicts
+
+#### CRD - CiliumNetworkPolicy, CiliumClusterwideNetworkPolicy
+
+* Extend the existing policy structure to include a new Authentication field
+
+```yaml
+apiVersion: "cilium.io/v2"
+kind: CiliumNetworkPolicy
+metadata:
+  name: "l3-egress-rule"
+spec:
+  endpointSelector:
+    matchLabels:
+      run: ssh-client
+  egress:
+  - toEndpoints:
+    - matchLabels:
+        run: sshd
+    auth:
+       enabled: true
+```
+
+* Semantics are:
+  * Apply the policy to pods with the label `run: ssh-client`
+  * Allow traffic to pods with the label `run: sshd`
+  * Require authentication for these connections
+  * Enable default deny
+* Initially, require `toEndpoints` field when the `auth` field is set
+  * Cannot provide mTLS with cluster-external peers
+  * Cannot provide mTLS with cluster entities
+* If there are two conflicting policies, one that requires authentication and one that does not require authentication, then authentication will be required
+* If there are two policies with conflicting authentication fields, reject the rule
+  * Log an error in the Cilium agent logs
+  * Emit a warning status to in-pod `cilium status` output
+  * Increment a metric, with tags
+* Gotchas / corner cases:
+  * Avoid requiring auth to kube-apiserver
+  * Avoid requiring auth to SPIRE
+
+
+### mTLS, SPIFFE, and SPIRE
+
+When building an mTLS solution, SPIFFE, the API for having workloads request an attested identity with a clear chain of custody, and SPIRE, the implementation of that API, are the state of the art. The SPIFFE project has spent a lot of time thinking deeply about how you can avoid common identity-sharing pitfalls like spoofing identities, getting identity details (commonly but not always X.509 TLS keypairs, but not always) to the workload security, and being able to easily rotate and revoke these identities when required.
```

Comment:
revokation is still in development:
https://github.com/spiffe/spire/issues/1934 with work being put into the local authority api mostly: https://github.com/spiffe/spire-api-sdk/pull/37

Reply:
rotation and revocation already exist on SPIRE (it is done automatically depending on authorities expirations).
That proposal is allow forcing that to happens  on-demand

Reply:
What does it mean to "automatically" do it? What is the responsibility of the SDK vs. Cilium agent code?

The datapath needs to react to the same revocation event somehow, so either the SDK exposes some event notification mechanism through the control plane to inform it, or we pre-build something like timers in and then each component can independently write their own logic to ensure everything is in sync. (Latter doesn't make sense to me for revocation cases, we *might* get away with it for rotation)

Reply:
The SPIRE server will, by default, re-key SVIDs a reasonable period of time before their expiry (I think the default is half their lifetime). The existing Cilium implementation will receive an "update" event including the new key when the SPIRE server is done (via the local SPIRE agent), and @meyskens is already working on handling that event in cilium/cilium#24300, and kicking off a re-auth and associated update of the auth table after that.

The linked proposal for SPIRE is to have a way to force revocation at any time, which would then generate a rekey event, which would end up in the Cilium agent via the SPIRE agent, and be processed in the same way.

Without a forced revocation event in SPIRE, it's really important to ensure that the certificate lifetime is short enough to limit the exposure window in the event of a keypair compromise.

Reply:
Thanks for the explanation, makes sense. The main reason I raise this is that again, if we assume that non-SPIRE implementations will eventually implement this CFP, it will be useful to know that the design assumes that the auth mechanism provides such "update events", so that is a key integration point that must be implemented.